std/json.zig

```diff
@@ -0,0 +1,1308 @@
+                std.debug.warn("[\n");
+
+                for (inner.toSliceConst()) |value| {
+                    if (not_first) {
+                        std.debug.warn(",\n");
+                    }
+                    not_first = true;
+                    padSpace(level + indent);
+                    value.toStringIndentLevel(indent, level + indent);
+                }
+                std.debug.warn("\n");
+                padSpace(level);
+                std.debug.warn("]");
+            },
+            Value.Object => |inner| {
+                var not_first = false;
+                std.debug.warn("{{\n");
+                var it = inner.iterator();
+
+                while (it.next()) |entry| {
+                    if (not_first) {
+                        std.debug.warn(",\n");
+                    }
+                    not_first = true;
+                    padSpace(level + indent);
+                    std.debug.warn("\"{}\": ", entry.key);
+                    entry.value.toStringIndentLevel(indent, level + indent);
+                }
+                std.debug.warn("\n");
+                padSpace(level);
+                std.debug.warn("}}");
+            },
+        }
+    }
+
+    fn padSpace(indent: usize) void {
+        var i: usize = 0;
+        while (i < indent) : (i += 1) {
+            std.debug.warn(" ");
+        }
+    }
+};
+
+// A non-stream JSON parser which constructs a tree of Value's.
+const JsonParser = struct {
+    allocator: &Allocator,
+    state: State,
+    copy_strings: bool,
+    // Stores parent nodes and un-combined Values.
+    // Worst case scenario we have nested key, values and so need two times the stack size.
+    stack: [2 * StreamingJsonParser.max_stack_size]Value,
+    stack_used: u16,
+
+    const State = enum {
+        ObjectKey,
+        ObjectValue,
+        ArrayValue,
+        Simple,
+    };
+
+    pub fn init(allocator: &Allocator, copy_strings: bool) JsonParser {
+        return JsonParser {
+            .allocator = allocator,
+            .state = State.Simple,
+            .copy_strings = copy_strings,
+            .stack = undefined,
+            .stack_used = 0,
+        };
+    }
+
+    pub fn reset(p: &JsonParser) void {
+        p.state = State.Simple;
+        p.stack_used = 0;
+    }
+
+    pub fn parse(p: &JsonParser, input: []const u8) !ValueTree {
+        var mp = StreamingJsonParser.init();
+
+        var arena = ArenaAllocator.init(p.allocator);
+        errdefer arena.deinit();
+
+        for (input) |c, i| {
+            var mt1: ?Token = undefined;
+            var mt2: ?Token = undefined;
+
+            try mp.feed(c, &mt1, &mt2);
+            if (mt1) |t1| {
+                try p.transition(&arena.allocator, input, i, t1);
+
+                if (mt2) |t2| {
+                    try p.transition(&arena.allocator, input, i, t2);
+                }
+            }
+        }
+
+        // Handle top-level lonely number values.
+        {
```

Comment:
This should be a separate function such as `mp.end(&mt1)`. The only token that is possible here would be to terminate a number value. We can never get multiple.